File: src/backend/app/models/job.py
from enum import Enum
from typing import Literal, Optional
from uuid import uuid4

from pydantic import BaseModel, Field


class JobStatus(str, Enum):
    pending = "pending"
    extracting_audio = "extracting_audio"
    transcribing = "transcribing"
    correcting = "correcting"
    translating = "translating"
    generating_subtitle = "generating_subtitle"
    done = "done"
    failed = "failed"


SubtitleMode = Literal["source", "target", "bilingual"]


class JobCreateOptions(BaseModel):
    source_language: str = "en"
    target_language: str = "zh"
    correct: bool = True
    asr_model: str = "paraformer-v2"
    llm_model: str = "default"
    subtitle_mode: SubtitleMode = "bilingual"


class JobRead(BaseModel):
    id: str = Field(default_factory=lambda: "job_{0}".format(uuid4().hex[:12]))
    status: JobStatus = JobStatus.pending
    progress: int = 0
    message: str = "Task is waiting to start."
    source_language: str = "en"
    target_language: str = "zh"
    correct: bool = True
    asr_model: str = "paraformer-v2"
    llm_model: str = "default"
    subtitle_mode: SubtitleMode = "bilingual"
    original_filename: Optional[str] = None
    input_extension: str = ".mp4"
    video_url: Optional[str] = None
    subtitle_url: Optional[str] = None
    srt_download_url: Optional[str] = None
    error: Optional[str] = None

File: src/frontend/src/types/job.ts
export type JobStatus =
  | "pending"
  | "extracting_audio"
  | "transcribing"
  | "correcting"
  | "translating"
  | "generating_subtitle"
  | "done"
  | "failed";

export type SubtitleMode = "source" | "target" | "bilingual";

export interface JobCreateOptions {
  source_language: string;
  target_language: string;
  correct: boolean;
  asr_model: string;
  llm_model: string;
  subtitle_mode: SubtitleMode;
}

export interface JobRead extends JobCreateOptions {
  id: string;
  status: JobStatus;
  progress: number;
  message: string;
  original_filename: string | null;
  input_extension: string;
  video_url: string | null;
  subtitle_url: string | null;
  srt_download_url: string | null;
  error: string | null;
}

export const statusLabel: Record<JobStatus, string> = {
  pending: "等待开始",
  extracting_audio: "提取音频中",
  transcribing: "语音识别中",
  correcting: "文本修正中",
  translating: "字幕翻译中",
  generating_subtitle: "生成字幕中",
  done: "已完成",
  failed: "失败",
};

export function isFinishedJob(job: JobRead): boolean {
  return job.status === "done" || job.status === "failed";
}

export function isRunningJob(job: JobRead): boolean {
  return !isFinishedJob(job) && job.status !== "pending";
}
